### ISSUE TITLE
Security: Generate and document GPG key(s)

### ISSUE THREAD

**Issue by hoffie** [jamulussoftware/jamulus#1495]
**Has this feature been discussed and generally agreed?**

#319 was about documenting the process and a contact address for reporting security issues. It was also discussed to provide a way to contact the project in encrypted form. This still needs to be done.

**Describe the solution you'd like**

Variants:
1. Document each developer's own GPG key (and address) on their team page (tbd)
2. Generate a key for team@ and share it.
Both variants have advantages and disadvantages, see https://github.com/jamulussoftware/jamuluswebsite/issues/319#issuecomment-803663521 and later comments.

**Comment by gene96817** [comment]
> 2. Generate a key for team@ and share it.

Sharing a key is generally considered a bad idea. If one member is compromised, the key has to be revoked for everyone. With individual keys, each member is accountable and when compromised, only one member is affected. In addition, if a member leaves the project, nothing special needs to be done. One of the benefits of GPG is that every message is signed and only the owner. In  can sign the message.

**Comment by DominikSchaller** [comment]
I think you linked the wrong issue.

**Comment by gilgongo** [comment]
In theory, I think we could automatically encrypt all incoming mail to the team address to each Jamulus maintainers' key. Would that help?

**Comment by gene96817** [comment]
That is the idea. In this way, you always know: 
- who sent the email, 
- you can verify the email was not modified

**Comment by gilgongo** [comment]
Well, you'd at least know it was was from the mail server for jamulus.io. Senders could also be confident that any reply had been read by the intended recipient, and if the recipient signed their reply the sender could check the fingerprint to the one on the website maybe?

**Comment by gene96817** [comment]
Many years ago, checking the fingerprint was a tedious, manual process.
Assuming you are using a mail reader that supports GPG (e.g. Outlook, Apple Mail), you would import the correspondent's public key. The mail reader does all the work. It can handle (1) encrypted mail or (2) signed mail (meaning the mail is in plaintext and it includes a signature to verity the mail has not been modified en route).

I don't know how many mail readers support GPG. I suggest you try using GPG for a few days. It is really convenient (compared to some 15+ years ago).

**Comment by gene96817** [comment]
I said it backwards. I don't know how may GPG plug-ins there are for mail readers.

**Comment by ann0see** [comment]
The GPG issue comes up again if we provide a .deb PPA.

**Comment by gilgongo** [comment]
In both cases, the problem to solve is that of having a single Jamulus Project signature/fingerprint, but with one or more signing keys so that we avoid having to share a private key in some way.  Is that right? I know that multiple keys can sign a single file, but then verification has to be done against the key(s) of whoever signs. Wonder how Debian does it?

**Comment by ann0see** [comment]
I think sharing the private (master) key would be the problem. 

Some maindev could also create a private master key sign our GPG keys and then delete the private key...

**Comment by gilgongo** [comment]
> This would mean that we can never sign new keys.

Although that's not necessarily a problem as new keys would be a very rare occurrence. Or have I misunderstood?

**Comment by ann0see** [comment]
True. But it's an inconvenience

**Comment by gilgongo** [comment]
Actually, reading this I think that might not be so bad:

https://mikeross.xyz/create-gpg-key-pair-with-subkeys/

We'd probably need to set the key to expire about once every couple of years anyway. But subkeys mean the master secret key can be kept offline.

**Comment by ann0see** [comment]
Ok. Do we have a consent here who stores the master key? I'd go with Volker if he wants to – as he's a neutral person.

**Comment by ann0see** [comment]
@jamulussoftware/maindevelopers both the macOS and debian signing PRs are now in. How do we continue here? I can of course create both, but is this OK? Probably, we'd go with a CA approach?

**Comment by pljones** [comment]
I'd rather find a CA we can agree on and all have access to.

**Comment by ann0see** [comment]
So you'd want a commercial CA?

**Comment by pljones** [comment]
Or https://letsencrypt.org/ might be a free option.

**Comment by ann0see** [comment]
@corrados Would you be willing to create a Certificate Authority as "center of trust" for the project? 

I think it would work as follows:

1. Volker creates a CA certificate
2. The Jamulus team sends their GPG keys or a Certificate signing request (think that's a CSR) to you which you sign
3. Someone creates a certificate for the debian repository which gets signed by the CA or a sub certificate
4. We deploy the GPG key for the repo

I think Volker as "neutral" party should have the "highest" certificate as he's the initial founder people must trust anyway and saves the certificate. 

He could then even sign a CA the team uses for production so that he's just a high level backup and source of trust if the team looses a certificate (this of course brings the problem of private key sharing).

I'm not too much into Cryptography, but I think this should work.

**Comment by ann0see** [comment]
The problem with a commercial CA is that they're mostly verifying domain names - I think.

**Comment by corrados** [comment]
> Would you be willing to create a Certificate Authority as "center of trust" for the project?

Yes, I can do it. But you have to tell me exactly how to do it since I have never done this and I do not want to make mistakes...

> I think Volker as "neutral" party should have the "highest" certificate as he's the initial founder people must trust anyway and saves the certificate.

That honors me :-).

**Comment by ann0see** [comment]
Thanks! I will try to find or write a step by step guide. 

I think the easiest way would be a Linux machine. Do you have access to one?

**Comment by ann0see** [comment]
After some research, I think what we are planning is not standard for GPG keys: https://unix.stackexchange.com/questions/401577/gpg-trust-like-an-ca-cert

macOS could use this approach - but it's not that important there as it would only be for non release builds.

So I'd suggest the following for the 

## Debian repository Key (Fixes: https://github.com/jamulussoftware/repos/issues/1)

1. Volker generates a GPG key https://www.redhat.com/sysadmin/creating-gpg-keypairs and stores it somewhere securely. The E-Mail needs to be determined, but is probably the standard one (@pljones what do you think?)
2. Volker adds the GPG key as repository secret (we'd need to give him access to the repo, but that's no problem. If he wants, he can remove himself later on)

This means that the key isn't shared.

## Concerning a key for the e-mail communication

we could use the same key, but I don't think that's the solution. I *think* that each maintainer should publish his (public) GPG key to e.g. his Github profile. We wouldn't have a shared key for the E-Mail address, but if something encrypted gets sent to it, the respective maintainer should forward the message to other developers.

## Concerning signing pre-release builds with a self signed certificate for macOS

we could use the CA approach, which I hoped would also work for the GPG keys. Someone, maybe Volker (?) would generate the CA on a Mac and sign certificate/s we'd then add to the CI

* I think the Debian key creation is actionable and needs an OK from @corrados and @pljones 
* The macOS CA creation is also actionable
* The E-Mail question still remains open

What do you think?

**Comment by pljones** [comment]
> I think the Debian key creation is actionable and needs an OK from @corrados and @pljones

OK by me.

**Comment by ann0see** [comment]
Ok. Great. Then I'll provide a step by step guide here:

Creation of GPG key for Debian Repo:
1. On a linux machine, in terminal run `gpg --full-generate-key`
2. Select RSA and RSA (1)
3. Select a RSA keylength of 4096 bit
4. Select a expration date of 3y - I believe that's ok
5. Confirm that the information is correct by typing y
6. As name provide `The Jamulus Development Team`
7. As E-Mail provide the E-Mail from here: https://github.com/jamulussoftware/jamulus/blob/main/SECURITY.md not pasted here for anti spam reasons
8. As Comment provide: GPG key for signing Jamulus releases generated by the official deb repository
9. Confirm
10. Provide a secure password (this should not be published)
11. Export the private key by executing `gpg --export-secret-keys -a -o jamulus_test.asc [id]` The id is a long Hex number. 
12.  Add the content of the exported file to https://github.com/jamulussoftware/jamulus/settings/secrets/actions repository secrets. It should be named GPG_PRIVATE_KEY: https://github.com/jamulussoftware/jamulus/blob/1736d2ad7276db575762f8e4d6f23fd32428f9b9/.github/workflows/autobuild.yml#L446
13. Kick off a beta build

**Comment by pljones** [comment]
> 10. Provide a secure password (this should not be published)

But make sure it's not lost..! :)

**Comment by corrados** [comment]
Ok, I've done it. What I have not yet understood is if I need to store anything else but the secure password. But I guess this is not the case since I added the private key to the Jamulus repo secrets and this should contain all the necessary information, right?

**Comment by ann0see** [comment]
I've pushed https://github.com/jamulussoftware/jamulus/releases/tag/r3_9_1nightly1 to test the deb repo.

**Comment by ann0see** [comment]
>  Is it in our bio

I would do that, yes.

**Comment by ann0see** [comment]
Ok. On my repo everything works as expected after some smaller modifications. @corrados could you please regenerate a key without password and upload it?

**Comment by corrados** [comment]
Hmm, if there is no password applied and the private key is uploaded to Github, anybody could create the key. Or have I misunderstood some important part?

**Comment by ann0see** [comment]
Yes. However if we did use a password, it would not be more secure as it needs to be stored next to the GPG key. If the key is leaked, leaking the password would be easy too.

The fact that you create the key was mostly question of trust.

**Comment by corrados** [comment]
I am still confused. When I created the first key according to your step-by-step guide, I exported a private key and uploaded it to Github. But there was no additional file which I stored locally on my hard drive. This was consistent to your comment above: "_you don't need to store anything_".

> Otherwise someone of the current development team would have direct access to the key.

Ah, so does it mean the secure area of Github where I uploaded the private key is only accessible by myself and no one  else? That would then make sense.

**Comment by ann0see** [comment]
> I am still confused. When I created the first key according to your step-by-step guide, I exported a private key and uploaded it to Github.

Yes that's correct. But I believe the key is still stored in the GPG keychain - I think that's the case for me at least.

> Ah, so does it mean the secure area of Github where I uploaded the private key is only accessible by myself and no one else? That would then make sense.

Yes. I think you can't even read it after you've added the content - only overwrite it.

However, GitHub actions can read it (of course this means that if the runner is compromised, the key and all other secrets like the SSH key hoffie added could be stolen).

**Comment by corrados** [comment]
Ok, I have created a new key without a password and updated the secret "GPG_PRIVATE_KEY" with the new key.

**Comment by ann0see** [comment]
Great! Thank you very much. I'll update the nightly.

**Comment by ann0see** [comment]
I've now also added a GPG key to my GitHub profile.

**Comment by ann0see** [comment]
@pljones what is outstanding here for 3.10.0? The macOS stuff? I think it's too late for today.

**Comment by pljones** [comment]
It wasn't marked as done, so I assumed there was something not done that needed to be done for 3.10.0.  I don't know what.

**Comment by ann0see** [comment]
No. MacOS self signing is not blocking.

**Comment by pljones** [comment]
OK, so can this be closed under 3.10.0?

**Comment by ann0see** [comment]
If every main dev has a key in their profile, probably yes?

**Comment by ann0see** [comment]
Also SECURITY.md needs to be updated

**Comment by ann0see** [comment]
@softins if you use Gnome on a Linux machine, you can generate one via a GUI. I think it's named Paasword or Keychain?

**Comment by pljones** [comment]
Are we going to go anywhere _at all_ with this, or should the issue just be closed?

**Comment by gilgongo** [comment]
> Are we going to go anywhere _at all_ with this, or should the issue just be closed? (It's been open over three years and gone nowhere.)

We do at least have our public keys in our GH profiles somewhere (see my GH profile), so in theory somebody could send one or all of us a secure message with that (and @softins as I think you asked about this?).

I lost track of the more ambitious plan to generate a key for team@ though. Seems rather complicated.

**Comment by ann0see** [comment]
I can update my key on my profile (It may have expired). But it shouldn't be a problem for at least all of us having one in our profile. 
Then not have a shared GPG key, but tell users/reporters to send a message to the team mail and maybe tell that an encrypted email is incoming. Then send the encrypted email to each person individually.

**Comment by gilgongo** [comment]
That would work, although one thing that occurs to me is that we could have, say, "security-reports@jamulus.io" which encrypted the mail to all our keys, then forwarded that to team@ perhaps (I'd need to give the mail account our public keys of course, and keep them up to date, but after that I think a procmail recipe could run `gpg` with `formail`). I could try and see if that works?

**Comment by pljones** [comment]
Wouldn't it be easier to follow if
- there was a security-reports@ key pair
- people sent encrypted to security-reports@ with the published security-reports@ public key
- security-reports@ had our public keys and forwarded securely using them

That way team@ doesn't get involved for secure communications, which could cut down unintentionally sending insecure messages.

**Comment by gilgongo** [comment]
Ah OK so senders just need to have the security-reports@ public key and don't need to find ours. Would be easier, yes. It would need its own key pair to encrypt to our keys in any case. But it would it need to add our keys to their encrypted message, no? Not sure how PGP does that, but I suppose it's possible (could decrypt on the server then immediately re-encrypt to us I maybe?).

Can't prevent people sending encrypted to team@ though of course. But I could try gold-plating things with a header sniffer that rejects encrypted emails (with a message) and vice versa for security-reports@ maybe :-)

**Comment by ann0see** [comment]
Decrypting on some server sounds bad practice!

**Comment by pljones** [comment]
Sharing a private key sounds pretty bad practice, too.  Those are the only two ways of doing it, though.

**Comment by gilgongo** [comment]
This turns out to be rather more difficult than expected, and not only because incoming mail could be ASCII in the body, PGP-MIMEed, or as an attachment (binary or ASCII). I've got it to forward plaintext emails to me encrypted at least, but that breaks DKIM and any DMARC stuff, so Gmail puts it into spam :-(

How's best to progress on this? Shall I do a PR with what I've got or something? I'm [adapting a script here](https://hoij.net/howto/pgp-encrypt-email-with-procmail-and-gpg/).

**Comment by pljones** [comment]
Dropping milestone until something more concrete turns up.
